tephra-core/src/main/java/org/apache/tephra/TransactionManager.java

```diff
@@ -1123,15 +1112,14 @@ private boolean doTruncateInvalidTxBefore(long time) throws InvalidTruncateTimeE
     }
     
     // Find all invalid transactions earlier than truncateWp
-    Set<Long> toTruncate = Sets.newHashSet();
-    for (long wp : invalid) {
-      // invalid list is sorted, hence can stop as soon as we reach a wp >= truncateWp
-      if (wp >= truncateWp) {
-        break;
+    LongSet toTruncate = new LongArraySet();
+    for (long wp : invalidTxList.toRawList()) {
```

Comment:
Unfortunately using the normal for-each loop means there would be boxing. To avoid boxing, you have to use the iterator directly

```java
LongIterator it = invalidTxList.toRawList().iterator();
while (it.hasNext()) {
  long wp = it.nextLong();
  ...
}
```